mypy/dmypy_server.py

```diff
@@ -486,8 +513,179 @@ def fine_grained_increment(self,
             fg_update_time=t2 - t1,
             files_changed=len(removed) + len(changed))
 
-        status = 1 if messages else 0
         self.previous_sources = sources
+        return messages
+
+    def fine_grained_increment_follow_imports(self, sources: List[BuildSource]) -> List[str]:
+        """Like fine_grained_increment, but follow imports."""
+
+        # TODO:
+        #  - file events
+        #  - search path updates
+        #  - logging
+
+        changed_paths = self.fswatcher.find_changed()
+
+        assert self.fine_grained_manager is not None
+        fine_grained_manager = self.fine_grained_manager
+        graph = fine_grained_manager.graph
+        manager = fine_grained_manager.manager
+
+        orig_modules = list(graph.keys())
+
+        self.update_sources(sources)
+
+        sources_set = {source.module for source in sources}
+
+        # Find changed modules reachable from roots (or in roots) already in graph.
+        seen, changed, removed, new_files = self.follow_imports(
+            sources, graph, set(), changed_paths, sources_set
+        )
+        sources.extend(new_files)
+
+        # Process changes directly reachable from roots.
+        messages = fine_grained_manager.update(changed, removed)
+
+        # Follow deps from changed modules (still within graph).
+        worklist = changed[:]
+        while worklist:
+            module = worklist.pop()
+            if module[0] not in graph:
+                continue
+            sources2 = self.direct_imports(module, graph)
+            seen, changed, removed, new_files = self.follow_imports(
+                sources2, graph, seen, changed_paths, sources_set
+            )
+            sources.extend(new_files)
+            self.update_sources(new_files)
+            messages = fine_grained_manager.update(changed, removed)
+            # TODO: Removed?
+            worklist.extend(changed)
+
+        for module_id, state in graph.items():
+            refresh_suppressed_submodules(module_id, state.path, fine_grained_manager.deps, graph,
+                                          self.fscache)
+
+        # There may be new files that became available, currently treated as
+        # suppressed imports. Process them.
+        seen_suppressed = set()  # type: Set[str]
+        while True:
+            # TODO: Merge seen and seen_suppressed?
+            new_suppressed, seen_suppressed = self.find_added_suppressed(
```

Comment:
This isn't really `new_suppressed`, right? This is stuff that is becoming *un*suppressed?

Reply:
Yes, this is confusing. I'll rename the variable.